workspaces/orchestrator/plugins/orchestrator/src/api/OrchestratorAuth.ts

```diff
@@ -13,86 +13,151 @@
+
+      if (!apiRef) {
+        throw new Error(
+          `Unknown custom auth provider API of id "${providerApiId}". The provider API id should match the ApiRef id.`,
+        );
       }
-      default: {
+
+      const api = apiHolder.get(apiRef);
+      if (!api || typeof api.getAccessToken !== 'function') {
         throw new Error(
-          `Schema contains unsupported authentication provider: ${tokenDescriptor.provider}. The supported providers are: github, gitlab and microsoft`,
+          `API with id "${providerApiId}" was found but does not implement the required getAccessToken method for authentication.`,
         );
       }
-    }
-    return {
-      token,
-      provider: tokenDescriptor.provider,
-    };
-  }
-
-  public async authenticate(
-    tokenDescriptors: AuthTokenDescriptor[],
-  ): Promise<Array<AuthToken>> {
-    const authTokens: Array<AuthToken> = [];
-    for (const tokenDescriptor of tokenDescriptors) {
-      authTokens.push(await this.getToken(tokenDescriptor));
-    }
-    return authTokens;
-  }
-}
+
+      return api;
+    },
+    [app, apiHolder],
+  );
+
+  const getToken = useCallback(
+    async (tokenDescriptor: AuthTokenDescriptor): Promise<AuthToken> => {
+      // First try built-in providers
+      let authApi: OAuthApi | (OAuthApi & OpenIdConnectApi) | undefined;
+
+      switch (tokenDescriptor.provider.toLocaleLowerCase('en-US')) {
+        case 'github':
+          authApi = githubAuthApi;
+          break;
+        case 'gitlab':
+          authApi = gitlabAuthApi;
+          break;
+        case 'microsoft':
+          authApi = microsoftAuthApi;
+          break;
+        default:
+          if (!tokenDescriptor.custonmProviderApiId) {
+            throw new Error(
+              `Custom authentication provider API id is required for provider: ${tokenDescriptor.provider}`,
+            );
+          }
+          authApi = await findCustomProvider(
+            tokenDescriptor.custonmProviderApiId,
+          );
+          break;
+      }
+
+      if (authApi) {
+        const token = await getProviderToken(authApi, tokenDescriptor);
+        return {
+          token,
+          provider: tokenDescriptor.provider,
+        };
+      }
+
+      throw new Error(
+        `Unsupported authentication provider: ${tokenDescriptor.provider}. The supported built-in providers are: github, gitlab and microsoft`,
+      );
+    },
+    [
+      githubAuthApi,
+      gitlabAuthApi,
+      microsoftAuthApi,
+      findCustomProvider,
+      getProviderToken,
+    ],
+  );
+
+  const authenticate = useCallback(
+    async (
+      tokenDescriptors: AuthTokenDescriptor[],
+    ): Promise<Array<AuthToken>> => {
+      if (!app) {
+        throw new Error('App context is required for authentication');
+      }
+
+      const authTokens: Array<AuthToken> = [];
+      for (const tokenDescriptor of tokenDescriptors) {
+        authTokens.push(await getToken(tokenDescriptor));
+      }
```

Comment:
Do we need to sequentially check them or is it possible to use `Promise.all`?

Reply:
It must be sequential because each getToken call potentially shows a pop up to the user, there can be multiple popups that the user needs to log in to